fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java

```diff
@@ -158,19 +157,30 @@ public Optional<QueryPredicate> toQueryPredicate(@Nonnull final CorrelationIdent
             final Value rightChild = it.next();
             final Set<CorrelationIdentifier> leftChildCorrelatedTo = leftChild.getCorrelatedTo();
             final Set<CorrelationIdentifier> rightChildCorrelatedTo = rightChild.getCorrelatedTo();
-            if (leftChildCorrelatedTo.equals(innermostAliasSet) && !rightChildCorrelatedTo.contains(innermostAlias)) {
-                final Object comparand = compileTimeEvalFn.apply(rightChild);
-                return comparand == null
-                        ? Optional.empty()
-                        : Optional.of(new ValuePredicate(leftChild, new Comparisons.SimpleComparison(comparisonType, comparand)));
-            } else if (rightChildCorrelatedTo.equals(innermostAliasSet) && !leftChildCorrelatedTo.contains(innermostAlias)) {
-                final Object comparand = compileTimeEvalFn.apply(leftChild);
-                return comparand == null
-                        ? Optional.empty()
-                        : Optional.of(new ValuePredicate(rightChild, new Comparisons.SimpleComparison(swapBinaryComparisonOperator(comparisonType), comparand)));
-            } else if (!rightChildCorrelatedTo.contains(innermostAlias) && !leftChildCorrelatedTo.contains(innermostAlias)) {
+
+            if (leftChildCorrelatedTo.isEmpty() && rightChildCorrelatedTo.isEmpty()) {
                 return tryBoxSelfAsConstantPredicate();
             }
+            if (rightChildCorrelatedTo.contains(innermostAlias) && !leftChildCorrelatedTo.contains(innermostAlias)) {
+                if (leftChildCorrelatedTo.isEmpty()) {
+                    final Object comparand = compileTimeEvalFn.apply(leftChild);
+                    return comparand == null
+                           ? Optional.empty()
+                           : Optional.of(new ValuePredicate(rightChild, new Comparisons.SimpleComparison(swapBinaryComparisonOperator(comparisonType), comparand)));
```

Comment:
Since `null` is a permissible value, we should return something else so the caller can distinguish this case.